mne/beamformer/_compute_beamformer.py

```diff
@@ -239,3 +239,187 @@ def _prepare_beamformer_input(info, forward, label, picks, pick_ori,
+                                     'by using reduce_rank=True.')
+
+                idx_max = eig_vals.argmax()
+                max_ori = eig_vecs[:, idx_max]
+                Wk[:] = np.dot(max_ori, Wk)
+                Gk = np.dot(Gk, max_ori)
+
+                # compute spatial filter for NAI or unit-noise-gain
+                tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                denom = np.sqrt(tmp)
+                Wk /= denom
+                if weight_norm == 'nai':
+                    Wk /= np.sqrt(noise)
+
+                is_free_ori = False
+
+            elif beamformer == 'dics':
+                # Compute the direction of max power
+                u, s, _ = np.linalg.svd(power.real)
+                max_ori = u[:, 0]
+
+                Wk[:] = np.dot(max_ori, Wk)
+
+        else:  # do vector beamformer
+            if beamformer == 'lcmv':
+                # compute the filters:
+                if is_free_ori:
+                    # Free source orientation
+                    Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
+                else:
+                    # Fixed source orientation
+                    Wk /= Ck
+
+                # handle noise normalization with free/normal source
+                # orientation:
+                if weight_norm == 'nai':
+                    raise NotImplementedError('Weight normalization with '
+                                              'neural activity index is not '
+                                              'implemented yet with free or '
+                                              'fixed orientation.')
+
+                elif weight_norm == 'unit-noise-gain':
+                    noise_norm = np.sum(Wk ** 2, axis=1)
+                    if is_free_ori:
+                        noise_norm = np.sum(noise_norm)
+                    noise_norm = np.sqrt(noise_norm)
+                    if noise_norm == 0.:
+                        noise_norm_inv = 0.  # avoid division by 0
+                    else:
+                        noise_norm_inv = 1. / noise_norm
+                    Wk[:] *= noise_norm_inv
+
+    # picking source orientation maximizing output source power
+    if pick_ori == 'max-power':
+        W = W[0::3]
+    elif pick_ori == 'normal':
+        W = W[2::3]
+        is_free_ori = False
+
+    if beamformer == 'dics':
+        if weight_norm == 'unit-noise-gain':
+            # Scale weights so that W @ I @ W.T == I
+            if pick_ori is None and n_orient > 1:
+                # Compute the norm for each set of 3 dipoles
+                W = W.reshape(-1, 3, W.shape[1])
+                norm = np.sqrt(np.sum(W ** 2, axis=(1, 2)))
+                W /= norm[:, np.newaxis, np.newaxis]
+                W = W.reshape(-1, W.shape[2])
+            else:
+                # Compute the norm for each dipole
+                norm = np.sqrt(np.sum(W ** 2, axis=1))
+                W /= norm[:, np.newaxis]
```

Comment:
I find this version readable. It's nice to do vector computations when possible.

Reply:
... also the LCMV code / tests can be a bit slow so it would be nice to figure out at some point where the slowdowns are and see if there are additional vectorization steps we could take.